salt/states/idem.py

```diff
@@ -0,0 +1,136 @@
+    SLSs = []
+    if tree:
+        sls_sources.append("file://{}".format(tree))
+    for sls in sources:
+        path = pathlib.Path(sls)
+        if path.is_file():
+            ref = str(path.stem if path.suffix == ".sls" else path.name)
+            SLSs.append(ref)
+            implied = "file://{}".format(path.parent)
+            if implied not in sls_sources:
+                sls_sources.append(implied)
+        else:
+            SLSs.append(sls)
+    return sls_sources, SLSs
+
+
+def state(
+    name,
+    sls,
+    acct_file=None,
+    acct_key=None,
+    acct_profile=None,
+    cache_dir=None,
+    render=None,
+    runtime=None,
+    test=False,
+    tree=None,
+):
+    """
+    Call an idem state through a salt state
+
+    sls
+        A list of idem sls files or sources
+
+    cache_dir
+        The location to use for the cache directory
+
+    render
+        The render pipe to use, this allows for the language to be specified (jinja|yaml)
+
+    runtime
+        Select which execution runtime to use (serial|parallel)
+
+    tree
+        The directory containing sls files
+
+    acct_file
+        Path to the acct file used in generating idem ctx parameters.
+        Defaults to the value in the ACCT_FILE environment variable.
+
+    acct_key
+        Key used to decrypt the acct file.
+        Defaults to the value in the ACCT_KEY environment variable.
+
+    acct_profile
+        Name of the profile to add to idem's ctx.acct parameter
+        Defaults to the value in the ACCT_PROFILE environment variable.
+
+    .. code-block:: yaml
+
+    cheese:
+      idem.state:
+        - runtime: parallel
+        - sls:
+          - idem_state.sls
+          - sls_source
+
+    :maturity:      new
+    :depends:       acct, pop, pop-config, idem
+    :platform:      all
+    """
+    hub = __utils__["idem.hub"]()
+
+    if isinstance(sls, str):
+        sls = [sls]
+
+    sls_sources, SLSs = _get_refs(sls, tree or hub.OPT.idem.tree)
+
+    coro = hub.idem.state.apply(
+        name=name,
+        sls_sources=sls_sources,
+        render=render or hub.OPT.idem.render,
+        runtime=runtime or hub.OPT.idem.runtime,
+        subs=["states"],
+        cache_dir=cache_dir or hub.OPT.idem.cache_dir,
+        sls=SLSs,
+        test=test,
+        acct_file=acct_file or hub.OPT.acct.acct_file,
+        acct_key=acct_key or hub.OPT.acct.acct_key,
+        acct_profile=acct_profile or hub.OPT.acct.acct_profile or "default",
+    )
+    hub.pop.Loop.run_until_complete(coro)
+
+    errors = hub.idem.RUNS[name]["errors"]
+    running = hub.idem.RUNS[name]["running"]
+    comments = "\n".join(
+        "{}[{}]: {}".format(v.get("name", k), v["result"], v["comment"])
+        for k, v in running.items()
+        if v.get("comment")
+    )
+    return {
+        "name": name,
+        "result": not errors,
+        "comment": errors or comments,
+        "changes": {k: v["changes"] for k, v in running.items() if v.get("changes")},
+    }
```

Comment:
I really want some feedback on this bit of code.  `hub.idem.state.apply` runs a whole set of idem idem states then the returns (which look exactly like salt state returns) are stored in `hub.idem.RUNS[name]["running"].  I toyed with a couple ways of returning all of those state returns and in the end decided to concatenate all the comments from all the runs with their idem-state-run name and result and then group all the changes on top of the salt state "changes" return.  Does anyone have any better ideas or does this look OK?

Reply:
An example of what shows up in  `hub.idem.RUNS[name]["running"]`:
```
{'test_|-state_name2_|-idem_test_|-none_without_changes': {'__run_num': 2,
                                                           'changes': {},
                                                           'comment': 'Success!',
                                                           'name': 'idem_test',
                                                           'result': None},
 'test_|-state_name_|-idem_test_|-none_without_changes': {'__run_num': 1,
                                                          'changes': {},
                                                          'comment': 'Success!',
                                                          'name': 'idem_test',
                                                          'result': None}}
```

Reply:
Maybe a way to do this is to modify states in general so that they can return a list of salt returns.  That way puppet, chef, and ansible calls could return each individual thing they ran as an individual return as well :thinking:

Reply:
Resolved by #57993